File: src/main/java/tafqeetj/converters/DecimalToWordsConverter.java
package tafqeetj.converters;

import tafqeetj.exceptions.NumberOutOfRangeException;
import tafqeetj.numbers.FractionalRatioNames;
import tafqeetj.numbers.NumberSign;

import java.math.BigDecimal;

class DecimalToWordsConverter {
    private final IntegerToWordsConverter converter = IntegerToWordsConverter.getInstance();

    private static DecimalToWordsConverter instance;

    private DecimalToWordsConverter() {
    }

    public synchronized static DecimalToWordsConverter getInstance() {
        if (instance == null)
            instance = new DecimalToWordsConverter();
        return instance;
    }

    public DecimalInWords convert(BigDecimal number) {
        if (TafqeetRangeChecker.isOutOfRange(number))
            throw new NumberOutOfRangeException(number);
        String[] numberLeftAndRight = number.toPlainString().split("\\.");
        if (numberLeftAndRight.length == 2)
            numberLeftAndRight[1] = removeTrailingZeroes(numberLeftAndRight[1]);
        return convertToDecimalInWords(numberLeftAndRight);
    }

    private String removeTrailingZeroes(String number) {
        int end = number.length();
        for (int i = number.length() - 1; i > 0; i--)
            if (number.charAt(i) == '0')
                end = i;
            else break;
        return number.substring(0, end);
    }

    private DecimalInWords convertToDecimalInWords(String[] numberLeftAndRight) {
        IntegerInWords leftNumber = convertLeftNumber(numberLeftAndRight[0]);
        IntegerInWords rightNumber = convertRightNumber((numberLeftAndRight.length == 2 ? numberLeftAndRight[1] : "0"));
        int fractionalLength = 0;
        if (numberLeftAndRight.length == 2)
            fractionalLength = numberLeftAndRight[1].length();
        return buildDecimalInWords(leftNumber, rightNumber, fractionalLength);
    }

    private IntegerInWords convertLeftNumber(String leftNumber) {
        IntegerInWords number;
        number = converter.convert(Long.parseLong(leftNumber));
        if (leftNumber.startsWith("-"))
            number.setSign(NumberSign.NEGATIVE);
        return number;
    }

    private IntegerInWords convertRightNumber(String rightNumber) {
        //يرسل الرقم يمين الفاصلة العشرية كسالب ليعامل معاملة المضاف إليه ويجر
        IntegerInWords number = converter.convert(Long.parseLong("-" + rightNumber));
        //يتم تعيين الإشارة الموجبة كي لا يتم وضع كلمة سالب قبل الرقم يمين الفاصلة العشرية
        number.setSign(NumberSign.POSITIVE);
        return number;
    }

    private DecimalInWords buildDecimalInWords(IntegerInWords leftNumber,
                                               IntegerInWords rightNumber, int fractionalLength) {
        DecimalInWords decimalInWords = new DecimalInWords();
        decimalInWords.setNumberLeftDecimalPoint(leftNumber);
        decimalInWords.setNumberRightDecimalPoint(rightNumber);
        if (!leftNumber.toString().equals(""))
            decimalInWords.setFractionalRatioName(FractionalRatioNames.getRatio(fractionalLength));
        return decimalInWords;
    }
}

File: src/main/kotlin/tech/moussa/tafqeetk/converters/DecimalToWordsConverter.kt
package tech.moussa.tafqeetk.converters

import tech.moussa.tafqeetk.exceptions.NumberOutOfRangeException
import tech.moussa.tafqeetk.numbers.FractionalRatioNames
import tech.moussa.tafqeetk.numbers.NumberSign
import java.math.BigDecimal

class DecimalToWordsConverter private constructor() {
    private val converter: IntegerToWordsConverter = IntegerToWordsConverter.getInstance()

    fun convert(number: BigDecimal): DecimalInWords {
        if (TafqeetRangeChecker.isOutOfRange(number)) {
            throw NumberOutOfRangeException(number)
        }
        val numberLeftAndRight = number.toPlainString().split("\\.".toRegex()).toTypedArray()
        if (numberLeftAndRight.size == 2) {
            numberLeftAndRight[1] = this.removeTrailingZeroes(numberLeftAndRight[1])
        }
        return this.convertToDecimalInWords(numberLeftAndRight)
    }

    fun convert(number: Double): DecimalInWords {
        if (TafqeetRangeChecker.isOutOfRange(number)) {
            throw NumberOutOfRangeException(number)
        }
        val numberLeftAndRight = number.toString().split("\\.".toRegex()).toTypedArray()
        if (numberLeftAndRight.size == 2) {
            numberLeftAndRight[1] = this.removeTrailingZeroes(numberLeftAndRight[1])
        }
        return this.convertToDecimalInWords(numberLeftAndRight)
    }

    private fun removeTrailingZeroes(number: String): String {
        var end = number.length
        for (i in number.length - 1 downTo 1) {
            end = if (number[i] == '0') {
                i
            } else {
                break
            }
        }
        return number.substring(0, end)
    }

    private fun convertToDecimalInWords(numberLeftAndRight: Array<String>): DecimalInWords {
        val leftNumber = this.convertLeftNumber(numberLeftAndRight[0])
        val rightNumber = this.convertRightNumber(
                if (numberLeftAndRight.size == 2) {
                    numberLeftAndRight[1]
                } else {
                    "0"
                }
        )
        var fractionalLength = 0
        if (numberLeftAndRight.size == 2) {
            fractionalLength = numberLeftAndRight[1].length
        }
        return this.buildDecimalInWords(leftNumber, rightNumber, fractionalLength)
    }

    private fun convertLeftNumber(leftNumber: String): IntegerInWords {
        val number: IntegerInWords = this.converter.convert(leftNumber.toLong())
        if (leftNumber.startsWith("-")) {
            number.setSign(NumberSign.NEGATIVE)
        }
        return number
    }

    private fun convertRightNumber(rightNumber: String): IntegerInWords {
        //يرسل الرقم يمين الفاصلة العشرية كسالب ليعامل معاملة المضاف إليه ويجر
        val number = this.converter.convert("-$rightNumber".toLong())
        //يتم تعيين الإشارة الموجبة كي لا يتم وضع كلمة سالب قبل الرقم يمين الفاصلة العشرية
        number.setSign(NumberSign.POSITIVE)
        return number
    }

    private fun buildDecimalInWords(
            leftNumber: IntegerInWords,
            rightNumber: IntegerInWords,
            fractionalLength: Int
    ): DecimalInWords {
        val decimalInWords = DecimalInWords()
        decimalInWords.setNumberLeftDecimalPoint(leftNumber)
        decimalInWords.setNumberRightDecimalPoint(rightNumber)
        if (leftNumber.toString() != "") {
            decimalInWords.setFractionalRatioName(FractionalRatioNames.getRatio(fractionalLength))
        }
        return decimalInWords
    }

    companion object {
        private var instance: DecimalToWordsConverter? = null

        @Synchronized
        fun getInstance(): DecimalToWordsConverter {
            if (DecimalToWordsConverter.instance == null) {
                DecimalToWordsConverter.instance = DecimalToWordsConverter()
            }
            return DecimalToWordsConverter.instance!!
        }
    }
}